File: scripts/listener.py
#!/usr/bin/env python

import rospy
import time
from frequency_test.msg import SuperAwesome, Freq


def callback(SuperAwesomeData):
    global start_time
    global prev_msg
    global nRecv # number of messages received for each time interval
    nRecv += 1

    current_time = time.time()
    dt = current_time - start_time
    
    current_msg = SuperAwesomeData.count
    time_interval = 3 # count every 3 seconds
    if dt > time_interval:
        start_time = current_time
        nSent = current_msg - prev_msg
        hz = nSent / time_interval # message receiving frequency 
        info_loss = 1 - nRecv / float(nSent) # message loss rate
        nRecv = 0
        prev_msg = current_msg
        freq_msg = Freq()
        freq_msg.freq = hz
        freq_msg.loss = info_loss
        pub.publish(freq_msg)


def listener():
    rospy.init_node('listener', anonymous=True)
    rospy.Subscriber('chatter', SuperAwesome, callback)
    rospy.spin()

if __name__ == '__main__':
    pub = rospy.Publisher('freq', Freq, queue_size=1)
    nRecv = 0
    prev_msg = 0
    start_time = time.time()
    listener()

File: scripts/listener.cpp
#include <time.h>
#include "ros/ros.h"
#include "frequency_test/Freq.h"
#include "frequency_test/SuperAwesome.h"

int64_t prev_msg = 0; 
double nRecv = 0; // number of received messages in each time interval
double start_time; 
ros::Publisher pub; 

void chatterCallback(const frequency_test::SuperAwesome::ConstPtr& SuperAwesomeData)
{
  double current_time = ros::Time::now().toSec();
  double dt = current_time - start_time;
  
  double time_interval = 3; // calculate message receiving frequency every 3 seconds
  nRecv++;
  if (dt > time_interval) {
    start_time = current_time;
    int64_t current_msg = SuperAwesomeData->count;
	int64_t nSent = current_msg - prev_msg;
    double hz = nSent / time_interval; // message receiving frequency
    double info_loss = 1 - nRecv / nSent; // message loss rate via transmition
    nRecv = 0;
    prev_msg = current_msg;
	frequency_test::Freq freq_msg;
    freq_msg.freq = hz;
    freq_msg.loss = info_loss;
    pub.publish(freq_msg);
  }
}

int main(int argc, char **argv)
{
  ros::init(argc, argv, "listener");
  ros::NodeHandle n;
  start_time = ros::Time::now().toSec();

  pub = n.advertise<frequency_test::Freq>("freq", 1000);

  ros::Subscriber sub = n.subscribe("chatter", 1000, chatterCallback);

  ros::spin();

  return 0;
}

